Update architecture overview (#1304)
vdaas/vald: docs/overview/data-flow.md

# Data Flow

> This section describes the data flow inside Vald and how Vald's vector indexes are stored.

Reviewer: ```suggestion
On this page, we describe the data flow inside Vald and how vector indexes are stored in the Vald Cluster.
```

> This is the most important part for the users to understand Vald.

Reviewer: ```suggestion
It will help you to understand what Vald does from users' requests.
```

> The below image is the overview of Vald's architecture.

Reviewer: ```suggestion
The below image is the basic Vald architecture.
```

<img src="../../assets/image/svg/Vald Architecture Overview.svg" />

> We will explain this image in the following sections.

Reviewer: ```suggestion
We will explain using this image in the following sections.
```

# Data Flow

On this page, we describe the data flow inside Vald and how vector indexes are stored in the Vald Cluster.
It will help you to understand what Vald does from users' requests.

The below image is the basic Vald architecture.

<img src="../../assets/image/svg/Vald Architecture Overview.svg" />

We will explain using this image in the following sections.

## Insert

Users can insert a vector to Vald cluster using Insert API. When the insert command is processed, Vald will intelligently insert the index into the most suitable Vald Agent(s) base on the memory usage of the Vald Agent and the node.

To obtain the memory usage of Vald Agent and the node, Vald Discoverer is required to rank which is the most suitable agent to perform the insert request by talking with the kube-apiserver.

To make the insert command effective, the `CreateIndex` instruction is required by the Vald Index Manager or Vald Agent itself to update the index to provide extremely high searching performance to users.

> Please note that only one embedding space is supported in a single Vald Cluster, if you want to support multiple embedded spaces, you may need to consider to deploy multiple Vald cluster to support this use case.

Reviewer: We should mention about the current Vald can handle only a single embedding space.

Author: I think I have already mentioned here 🤔
`Please note that only one embedding space is supported in a single Vald Cluster`

# Data Flow

On this page, we describe the data flow inside Vald and how vector indexes are stored in the Vald Cluster.
It will help you to understand what Vald does from users' requests.

The below image is the basic Vald architecture.

<img src="../../assets/image/svg/Vald Architecture Overview.svg" />

We will explain using this image in the following sections.
- [Insert](#Insert)
- [Search](#Search)

> - [Insert](#Update)

Author: ```suggestion
- [Update](#Update)
```

- [Upsert](#Upsert)
- [Delete](#Delete)

## Insert

Users can insert a vector to Vald cluster using Insert API. When the insert command is processed, Vald will intelligently insert the index into the most suitable Vald Agent(s) base on the memory usage of the Vald Agent and the node.

To obtain the memory usage of Vald Agent and the node, Vald Discoverer is required to rank which is the most suitable agent to perform the insert request by talking with the kube-apiserver.

To make the insert command effective, the `CreateIndex` instruction is required by the Vald Index Manager or Vald Agent itself to update the index to provide extremely high searching performance to users.

Please note that only one embedding space is supported in a single Vald Cluster, if you want to support multiple embedded spaces, you may need to consider to deploy multiple Vald cluster to support this use case.

<img src="../../assets/docs/overview/insert_flow_v2.png" />

When the user inserts data into Vald:

> 1. Vald Ingress receives the request from the user. The request includes the vector and the vector ID.

Author: ```suggestion
1. Vald Ingress receives the request from the user. The request includes the vector and the vector ID. The vector ID is the user-defined unique ID for each vector.
```

2. Vald Ingress will forward the request to the Vald LB Gateway to process the request. Vald LB Gateway will determine which Vald Agent(s) to process the request based on the resource usage of the nodes and pods, and the number of vector replicas.

> 3. Vald LB Gateway will forward the UUID and the vector data to the selected Vald Agents in parallel. Vald Agent will insert the vector and UUID in an on-memory vector queue. A vector queue will be committed to an ANN graph index by a `CreateIndex` instruction executed by the Vald Index Manager.

Author: ```suggestion
3. Vald LB Gateway will forward the generated UUID and the vector data to the selected Vald Agents in parallel. Vald Agent will insert the vector and UUID in an on-memory vector queue. A vector queue will be committed to an ANN graph index by a `CreateIndex` instruction executed by the Vald Index Manager.
```

> 3. Vald LB Gateway will forward the UUID and the vector data to the selected Vald Agents in parallel. Vald Agent will insert the vector and UUID in an on-memory vector queue. A vector queue will be committed to an ANN graph index by a `CreateIndex` instruction executed by the Vald Index Manager.

Reviewer: https://github.com/vdaas/vald/pull/1304#discussion_r665047029

> 3. Vald LB Gateway will forward the UUID and the vector data to the selected Vald Agents in parallel. Vald Agent will insert the vector and UUID in an on-memory vector queue. A vector queue will be committed to an ANN graph index by a `CreateIndex` instruction executed by the Vald Index Manager.

Author: ```suggestion
3. VVald LB Gateway will generate the UUID, and forward the generated UUID and the vector data to the selected Vald Agents in parallel. Vald Agent will insert the vector and UUID in an on-memory vector queue. A vector queue will be committed to an ANN graph index by a `CreateIndex` instruction executed by the Vald Index Manager.
```

2. Vald Ingress will forward the request to the Vald LB Gateway to process the request.
3. Vald LB Gateway will forward the request to all Vald Agents in parallel. Each Vald Agent will search the _k_ nearest neighbor vectors in an on memory graph index.
4. Vald Agent returns the searching result to the Vald LB Gateway. The searching result includes the UUID, the vector distance, and the vector. The number of the result will be the same as requested.
5. Vald LB Gateway will aggregate all searching results from all Vald Agents, rank the result by the vector distance, and return the ranked result to the Vald Ingress.

## Update

Users can update a vector by sending an update request to Vald. Vald will perform delete and insert requests to perform a single update command.

<img src="../../assets/docs/overview/update_flow_v2.png" />

When the user updates a vector from Vald:

1. Vald Ingress receives the request from the user. The request includes the existing vector ID(s) and the new vector(s) to be updated.
2. Vald Ingress will forward the request to the Vald LB Gateway to process the request.

> 3. Vald LB Gateway will broadcast the delete request with UUID(s) to the Vald Agents. Each Vald Agent will delete the vector data and the metadata if the corresponding UUID(s) is found in the in-memory graph index.

Author: ```suggestion
3. Vald LB Gateway will broadcast the delete request with the generated UUID(s) to the Vald Agents. Each Vald Agent will delete the vector data and the metadata if the corresponding UUID(s) is found in the in-memory graph index.
```

> 3. Vald LB Gateway will broadcast the delete request with UUID(s) to the Vald Agents. Each Vald Agent will delete the vector data and the metadata if the corresponding UUID(s) is found in the in-memory graph index.

Reviewer: I meant when the UUID is generated?

> 3. Vald LB Gateway will broadcast the delete request with UUID(s) to the Vald Agents. Each Vald Agent will delete the vector data and the metadata if the corresponding UUID(s) is found in the in-memory graph index.

Author: ```suggestion
3. Vald LB Gateway will generate the UUID(s), and broadcast the delete request with the generated UUID(s) to the Vald Agents. Each Vald Agent will delete the vector data and the metadata if the corresponding UUID(s) is found in the in-memory graph index.
```
